File: core/src/main/java/com/zeyad/rxredux/core/redux/UIObserver.java
package com.zeyad.rxredux.core.redux;

import android.arch.lifecycle.Observer;
import android.util.Log;

import io.reactivex.annotations.NonNull;

/**
 * @author by ZIaDo on 11/11/17.
 */
public class UIObserver<V extends LoadDataView<S>, S> implements Observer<UIModel<S>> {
    @NonNull
    private final V view;
    @NonNull
    private final ErrorMessageFactory errorMessageFactory;

    public UIObserver(@NonNull V view, @NonNull ErrorMessageFactory errorMessageFactory) {
        this.view = view;
        this.errorMessageFactory = errorMessageFactory;
    }

    @Override
    public void onChanged(@NonNull UIModel<S> uiModel) {
        Log.d("onNext", "UIModel: " + uiModel.toString());
        boolean loading = uiModel.isLoading();
        view.toggleViews(loading);
        if (!loading) {
            if (uiModel.isSuccessful()) {
                S bundle = uiModel.getBundle();
                view.setState(bundle);
                view.renderSuccessState(bundle);
            } else {
                Throwable error = uiModel.getThrowable();
                if (error != null) {
                    Log.e("UIObserver", "onChanged", error);
                    view.showError(errorMessageFactory.getErrorMessage(error));
                }
            }
        }
    }
}

File: core/src/main/java/com/zeyad/rxredux/core/UIObserver.kt
package com.zeyad.rxredux.core

import android.arch.lifecycle.Observer
import android.util.Log

/**
 * @author ZIaDo on 2/27/18.
 */
class UIObserver<V : LoadDataView<S>, S>(private val view: V, private val errorMessageFactory: ErrorMessageFactory) :
        Observer<UIModel<S>> {
    override fun onChanged(uiModel: UIModel<S>?) {
        Log.d("onNext", "UIModel: " + uiModel.toString())
        val loading = uiModel?.isLoading
        view.toggleViews(loading!!)
        if (loading) {
            if (uiModel.isSuccessful) {
                val bundle = uiModel.getBundle()
                view.setState(bundle!!)
                view.renderSuccessState(bundle)
            } else {
                val error = uiModel.throwable
                if (error != null) {
                    Log.e("UIObserver", "onChanged", error)
                    view.showError(errorMessageFactory.getErrorMessage(error))
                }
            }
        }
    }
}